docs/spec.md

```diff
@@ -0,0 +1,269 @@
+
+We might want to consider enforcing a minimum set of properties all resources should satisfy. For example:
+
+- toString: return the string representation of a resource.
+- asValue: return the (non resource) calc value for a resource. This would be equivalent to dereferencing for cells if we treat cells as resources.
+
+[JW-NOTE: This level of abstraction comes at a price that we should consider. Do we really want or need to support different evaluation workflows?]
+
+A `Host` is any object that has a `notify` method. The function takes a resource and a property and indicates to the host that the property of the given resource has changed. A couple of design considerations about the type of `notify`. One option is to just make `notify` set some dirty bit so that the resource can re-evaluate its formula:
+
+- `notify: () => void;`
+
+The reason for not choosing this approach is because it rules out true reactive or observable to style programming where the `Host` is effectively an `Observer`. Another approach would be to pass the value in the notification so that `Host` really is an observer of values for that property:
+
+- `notify: (resource: Resource, property: string, value: CalcValue) => void;`
+
+A reason for not choosing this approach would be because it couples invalidation of a resource with its re-computation. If I want to dirty a cell and notify its dependents I must now recompute the cell first to obtain the new value. This may be expensive and the host may not even need the value right now. Additionally, it is possible to support this observer style signature using the currently proposed one with a wrapper function:
+
+```ts
+interface Observer {
+  notify: (resource: Resource, property: string, value: CalcValue) => void;
+}
+
+interface Host {
+  notify: (resource: Resource, property) => void;
+}
+
+function observerToHost(observer: Observer): Host {
+  const host = {
+    notify(resource: Resource, property: string) {
+      resource.request(
+        host,
+        property,
+        value => observer.notify(resource, property, value),
+        () => {
+          throw "err";
+        }
+      );
+    }
+  };
+  return host;
+}
+```
+
+[JW-NOTE: The Resource/Host model is definitely on open design question so any guidance or alternate design feedback would be great here. In particular, I'm unsure of the type for `Host`. In many ways I like the simplicity of `{ notify: (resource: Resource, property: string, value: CalcValue) => void; }` but I'm unsure about having to pass the value down. How useful is that value, given that we probably need to go off and request all the other dependencies to do a recalc anyway. We can omit the value and encode the observer-stye signature, but at the cost of making an additional request to the resource.]
+
```

Comment:
I really like that idea!

> Producers also kept a second list of consumers who didn't understand their invalidation protocol. These consumers were just notified that "something changed".

We should try to aim to make the experience as good as possible for this second class of consumers too. At least in my head, there are lots of scenarios where consumers are plugging into produces they don't statically know about (and it will be the user driving the consumer through formula authoring).

I wonder if there are some general consumer interfaces that we could capture. Those that just recalc on any change (dirty bit), those that want the specific values, etc. Not sure how much that gives us though.